### PR TITLE
SPM functionality

### PR DESCRIPTION
**Pull request by rijobro**
If Matlab is available in SIRF, this will enable registration and resampling via SPM.
Registration currently in C++, needs to be exported to Python and Matlab.

### CODE REVIEW

File: src/Registration/CMakeLists.txt [hunk 1]
```diff
@@ -18,6 +18,13 @@
 
 # Have we got SPM12? (only works with Matlab)
 if (BUILD_MATLAB)
+	# if on linux, engine is in funny place
+	if("${CMAKE_SYSTEM}" MATCHES "Linux")
+		list(APPEND Matlab_LIBRARIES 
+			"${Matlab_ROOT_DIR}/extern/bin/glnxa64/libMatlabEngine.so" 
+			"${Matlab_ROOT_DIR}/extern/bin/glnxa64/libMatlabDataArray.so")
```
Comment (KrisThielemans) [inline src/Registration/CMakeLists.txt]: I'd need to see solid evidence of this. It'd be a CMake bug, which is likely version  specific.

Also, I seem to recall that "Linux" will match OSX as well (not sure)

File: CMakeLists.txt [hunk 2]
```diff
@@ -66,6 +66,7 @@ if (APPLE)
   if (SHARED_LIBS_ABS_PATH)
     # Set install_name_dir as the absolute path to install_prefix/lib
     GET_FILENAME_COMPONENT(CMAKE_INSTALL_NAME_DIR ${CMAKE_INSTALL_PREFIX}/lib REALPATH)
+    set(CMAKE_INSTALL_RPATH_USE_LINK_PATH TRUE)
```
Comment (KrisThielemans) [inline CMakeLists.txt]: a bit surprising to see this in this PR... would need to be listed in `CHANGES.md` I feel. It's going to affect people. Better to put in a separate PR?

File: cmake/FindSPM12.cmake [hunk 3]
```diff
@@ -0,0 +1,29 @@
+# Copyright 2019 University College London
```
Comment (KrisThielemans) [inline cmake/FindSPM12.cmake]: might have to be 2020!

File: src/Registration/cReg/include/sirf/Reg/NiftiBasedRegistration.h [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+#include "sirf/Reg/Registration.h"
+
+namespace sirf {
+
+/// Forward declarations
+template<class dataType> class NiftiImageData3D;
+
+/*!
+\ingroup Registration
+\brief Base class for all NiftyReg registrations.
```
Comment (KrisThielemans) [inline src/Registration/cReg/include/sirf/Reg/NiftiBasedRegistration.h]: confused between Nifti and Nifty

Comment (rijobro) [reply src/Registration/cReg/include/sirf/Reg/NiftiBasedRegistration.h]: Good spot, apologies.

File: CMakeLists.txt (hunk 2 above)
Comment (rijobro) [reply CMakeLists.txt]: deleted

File: src/Registration/CMakeLists.txt (hunk 1 above)
Comment (rijobro) [reply src/Registration/CMakeLists.txt]: > I'd need to see solid evidence of this. It'd be a CMake bug, which is likely version specific.

Sure. I wanted to do `find_package(Matlab COMPONENTS QUIET ENG_LIBRARY)` or similar, but couldn't get it to return true. I used info [here](https://uk.mathworks.com/help/matlab/matlab_external/build-c-engine-programs.html) and [here](https://uk.mathworks.com/matlabcentral/answers/472865-calling-matlab-functions-from-c-application-built-with-cmake-in-visual-studio#answer_387233), and the first link doesn't seem to mention anything about the location changing for different Matlab versions. As per the second link, we should probably add in the Windows location, too, but I can't test that.

> Also, I seem to recall that "Linux" will match OSX as well (not sure)

`if (UNIX)` will return true for OSX, whereas `if("${CMAKE_SYSTEM}" MATCHES "Linux")` won't (tested it).

Comment (KrisThielemans) [reply src/Registration/CMakeLists.txt]: pretty weird. 
https://github.com/Kitware/CMake/blob/451360c72914d28cc10ba36d2215aed756e568e5/Modules/FindMatlab.cmake#L1681-L1691
seems to look for it in the same place as the other libraries.
I see that it goes only looking for this for recent versions of MATLAB
https://github.com/Kitware/CMake/blob/451360c72914d28cc10ba36d2215aed756e568e5/Modules/FindMatlab.cmake#L1551-L1555

so what MATLAB version do you have (and CMake version).

Note that since CMake 3.17.0 you can debug `find_library`, see https://gitlab.kitware.com/cmake/cmake/merge_requests/3935

Comment (rijobro) [reply src/Registration/CMakeLists.txt]: I have Matlab2019a, but cmake 3.13.4 (current version via `apt install`).

Comment (rijobro) [reply src/Registration/CMakeLists.txt]: After the `FIND_PACKAGE(Matlab QUIET COMPONENTS)` on the linux machine, the following cmake variables contain the matlab keyword:
```
-- MATLAB_DEST_DIR=
-- MATLAB_FOUND=TRUE
-- MATLAB_INCLUDE_DIR_TO_LOOK=/usr/local/MATLAB/R2019a/extern/include
-- MATLAB_ROOT=/usr/local/MATLAB/R2019a
-- MATLAB_VERSIONS_MAPPING=R2018b=9.5;R2018a=9.4;R2017b=9.3;R2017a=9.2;R2016b=9.1;R2016a=9.0;R2015b=8.6;R2015a=8.5;R2014b=8.4;R2014a=8.3;R2013b=8.2;R2013a=8.1;R2012b=8.0;R2012a=7.14;R2011b=7.13;R2011a=7.12;R2010b=7.11
-- Matlab_BINARIES_DIR=/usr/local/MATLAB/R2019a/bin/glnxa64
-- Matlab_DATAARRAY_LIBRARY=
-- Matlab_DIR=Matlab_DIR-NOTFOUND
-- Matlab_ENGINE_LIBRARY=
-- Matlab_ENG_LIBRARY=
-- Matlab_EXTERN_LIBRARY_DIR=/usr/local/MATLAB/R2019a/extern/lib/glnxa64
-- Matlab_FOUND=TRUE
-- Matlab_FOUND=TRUE
-- Matlab_INCLUDE_DIRS=/usr/local/MATLAB/R2019a/extern/include
-- Matlab_LIBRARIES=/usr/local/MATLAB/R2019a/bin/glnxa64/libmex.so
-- Matlab_MAIN_PROGRAM=/usr/local/MATLAB/R2019a/bin/matlab
-- Matlab_MAIN_PROGRAM_FOUND=TRUE
-- Matlab_MAT_LIBRARY=
-- Matlab_MEXEXTENSIONS_PROG=
-- Matlab_MEX_EXTENSION=mexa64
-- Matlab_MEX_LIBRARY=/usr/local/MATLAB/R2019a/bin/glnxa64/libmex.so
-- Matlab_MX_LIBRARY=
-- Matlab_Or_MCR=UNKNOWN
-- Matlab_PROG_VERSION_STRING_AUTO_DETECT=/usr/local/MATLAB/R2019a/bin/matlab
-- Matlab_ROOT_DIR=/usr/local/MATLAB/R2019a
-- Matlab_ROOT_DIR=/usr/local/MATLAB/R2019a
-- Matlab_ROOT_DIR_LAST_CACHED=/usr/local/MATLAB/R2019a
```

So I have no idea what the `# The MatlabEngine library is required for R2018a+` comment means, since its not finding it.

Comment (KrisThielemans) [reply src/Registration/CMakeLists.txt]: I note `MATLAB_VERSIONS_MAPPING` doesn't include ¬R2019a` for your CMake. Either upgrade CMake, or use `MATLAB_ADDITIONAL_VERSION`. Maybe that'll fix it
